File: app/src/main/java/com/theandroiddev/mywins/InsertSuccessActivity.java
package com.theandroiddev.mywins;

import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.DisplayMetrics;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.theandroiddev.mywins.data.models.Success;
import com.theandroiddev.mywins.editsuccess.EditDescriptionActivity;
import com.theandroiddev.mywins.utils.DateHelper;
import com.theandroiddev.mywins.utils.DrawableSelector;

import static com.theandroiddev.mywins.utils.Constants.DATE_ENDED_EMPTY;
import static com.theandroiddev.mywins.utils.Constants.DATE_STARTED_EMPTY;
import static com.theandroiddev.mywins.utils.Constants.EXTRA_DESCRIPTION;
import static com.theandroiddev.mywins.utils.Constants.EXTRA_INSERT_SUCCESS_ITEM;
import static com.theandroiddev.mywins.utils.Constants.REQUEST_CODE_DESCRIPTION;
import static com.theandroiddev.mywins.utils.Constants.dummyImportanceDefault;

public class InsertSuccessActivity extends AppCompatActivity implements View.OnClickListener {


    private static final String TAG = "asd";
    TextView categoryTv, importanceTv, dateStartedTv, dateEndedTv;
    EditText titleEt, description_et;
    ImageView categoryIv, dateStartedIv, dateEndedIv, descriptionIv, importance1Iv, importance2Iv, importance3Iv, importance4Iv;
    Button addBtn;
    DrawableSelector drawableSelector;
    DateHelper dateHelper;
    private DisplayMetrics displayMetrics;
    private int accentColor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_insert_success);

        initWindow();
        initViews();
        setFields();

    }

    private void setFields() {

        String category = getIntent().getExtras().getString("categoryName", "other");
        categoryTv.setText(category);

        drawableSelector.selectCategoryImage(categoryIv, category, categoryTv);
        drawableSelector.setImportance(dummyImportanceDefault, importanceTv, importance1Iv, importance2Iv, importance3Iv, importance4Iv);

    }

    private void initViews() {

        drawableSelector = new DrawableSelector(this);
        dateHelper = new DateHelper(this);

        categoryTv = findViewById(R.id.insert_category_tv);
        importanceTv = findViewById(R.id.insert_importance_tv);
        titleEt = findViewById(R.id.instert_title_et);
        dateStartedTv = findViewById(R.id.insert_date_started_tv);
        dateEndedTv = findViewById(R.id.insert_date_ended_tv);
        description_et = findViewById(R.id.insert_description_et);
        categoryIv = findViewById(R.id.insert_category_iv);
        dateStartedIv = findViewById(R.id.insert_date_started_iv);
        dateEndedIv = findViewById(R.id.insert_date_ended_iv);
        descriptionIv = findViewById(R.id.insert_description_iv);
        importance1Iv = findViewById(R.id.insert_importance_iv_1);
        importance2Iv = findViewById(R.id.insert_importance_iv_2);
        importance3Iv = findViewById(R.id.insert_importance_iv_3);
        importance4Iv = findViewById(R.id.insert_importance_iv_4);
        addBtn = findViewById(R.id.insert_add_btn);


        accentColor = ResourcesCompat.getColor(getResources(), R.color.accent, null);

        dateStartedIv.setColorFilter(accentColor);
        descriptionIv.setColorFilter(accentColor);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            addBtn.setBackgroundTintList(getResources().getColorStateList(R.color.accent, null));
        } else {
            addBtn.setBackgroundColor(ContextCompat.getColor(this, R.color.accent));
        }

        dateEndedIv.setColorFilter(accentColor);
        descriptionIv.setColorFilter(accentColor);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            addBtn.setBackgroundTintList(getResources().getColorStateList(R.color.accent, null));
        } else {
            addBtn.setBackgroundColor(ContextCompat.getColor(this, R.color.accent));
        }

        importance1Iv.setOnClickListener(this);
        importance2Iv.setOnClickListener(this);
        importance3Iv.setOnClickListener(this);
        importance4Iv.setOnClickListener(this);
        dateStartedTv.setOnClickListener(this);
        dateEndedTv.setOnClickListener(this);
        dateStartedIv.setOnClickListener(this);
        dateEndedIv.setOnClickListener(this);
        descriptionIv.setOnClickListener(this);
        addBtn.setOnClickListener(this);

    }

    private void initWindow() {

        displayMetrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        int width = displayMetrics.widthPixels;
        int height = displayMetrics.heightPixels;
        getWindow().setLayout((int) (width * .9), (int) (height * .8));

    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.insert_importance_iv_1:
                drawableSelector.setSmallImportance(importanceTv, importance1Iv, importance2Iv, importance3Iv, importance4Iv);
                break;
            case R.id.insert_importance_iv_2:
                drawableSelector.setMediumImportance(importanceTv, importance1Iv, importance2Iv, importance3Iv, importance4Iv);
                break;
            case R.id.insert_importance_iv_3:
                drawableSelector.setBigImportance(importanceTv, importance1Iv, importance2Iv, importance3Iv, importance4Iv);
                break;
            case R.id.insert_importance_iv_4:
                drawableSelector.setHugeImportance(importanceTv, importance1Iv, importance2Iv, importance3Iv, importance4Iv);
                break;
            case R.id.insert_date_started_iv:
                dateHelper.setDate(DATE_STARTED_EMPTY, dateStartedTv, dateEndedTv);
                break;
            case R.id.insert_date_ended_iv:
                dateHelper.setDate(DATE_ENDED_EMPTY, dateStartedTv, dateEndedTv);
                break;
            case R.id.insert_date_started_tv:
                dateHelper.setDate(DATE_STARTED_EMPTY, dateStartedTv, dateEndedTv);
                break;
            case R.id.insert_date_ended_tv:
                dateHelper.setDate(DATE_ENDED_EMPTY, dateStartedTv, dateEndedTv);
                break;
            case R.id.insert_description_iv:
                setDesc();
                break;
            case R.id.insert_add_btn:
                insertSuccess();
                break;

            default:
                break;
        }
    }

    private void insertSuccess() {

        if (dateHelper.validateData(titleEt, dateStartedTv, dateEndedTv)) {

            String dateStarted = dateHelper.checkBlankDate(dateStartedTv.getText().toString());
            String dateEnded = dateHelper.checkBlankDate(dateEndedTv.getText().toString());

            sendData(dateStarted, dateEnded);
        }
    }

    private void sendData(String dateStarted, String dateEnded) {

        Intent returnIntent = new Intent();

        Success s = new Success(titleEt.getText().toString(), categoryTv.getText().toString(),
                drawableSelector.getImportance(importanceTv.getText().toString()), description_et.getText().toString(), dateStarted, dateEnded);

        //s.setId(UUID.randomUUID().toString());
        //Log.e(TAG, "sendData: uuid" + s.getId() );


        returnIntent.putExtra(EXTRA_INSERT_SUCCESS_ITEM, s);
        setResult(Activity.RESULT_OK, returnIntent);
        finish();
    }


    private void setDesc() {

        Intent intent = new Intent(InsertSuccessActivity.this, EditDescriptionActivity.class);
        intent.putExtra("description", description_et.getText().toString());
        startActivityForResult(intent, REQUEST_CODE_DESCRIPTION);

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REQUEST_CODE_DESCRIPTION) {
            if (resultCode == RESULT_OK) {
                String description = data.getExtras().getString(EXTRA_DESCRIPTION);
                description_et.setText(description);

            }
        }
    }
}

File: app/src/main/java/com/theandroiddev/mywins/presentation/insert_success/InsertSuccessActivity.kt
package com.theandroiddev.mywins.presentation.insert_success

import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v4.content.res.ResourcesCompat
import android.util.DisplayMetrics
import android.view.View
import com.theandroiddev.mywins.R
import com.theandroiddev.mywins.core.mvp.MvpDaggerAppCompatActivity
import com.theandroiddev.mywins.presentation.edit_success.EditDescriptionActivity
import com.theandroiddev.mywins.presentation.successes.SuccessModel
import com.theandroiddev.mywins.utils.Constants
import com.theandroiddev.mywins.utils.Constants.Companion.Category
import com.theandroiddev.mywins.utils.Constants.Companion.Importance
import com.theandroiddev.mywins.utils.DateHelper
import com.theandroiddev.mywins.utils.DrawableSelector
import kotlinx.android.synthetic.main.activity_insert_success.*

class InsertSuccessActivity : MvpDaggerAppCompatActivity<InsertSuccessView,
        InsertSuccessBundle, InsertSuccessPresenter>(), InsertSuccessView, View.OnClickListener {


    var drawableSelector = DrawableSelector(this)
    var dateHelper = DateHelper(this)
    private var displayMetrics: DisplayMetrics? = null
    private var accentColor: Int = 0

    private var currentCategory: Category = Category.NONE
    private var currentImportance: Importance = Importance.NONE

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_insert_success)

        initWindow()
        initViews()
        presenter.onAfterCreate()

    }

    override fun onBackPressed() {
        val intentParent = intent
        setResult(Activity.RESULT_CANCELED, intentParent)
        finish()
    }

    private fun initViews() {

        drawableSelector = DrawableSelector(this)
        dateHelper = DateHelper(this)

        accentColor = ResourcesCompat.getColor(resources, R.color.accent, null)

        insert_date_started_iv.setColorFilter(accentColor)
        insert_description_iv.setColorFilter(accentColor)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            insert_add_btn.backgroundTintList = resources.getColorStateList(R.color.accent, null)
        } else {
            insert_add_btn.setBackgroundColor(ContextCompat.getColor(this, R.color.accent))
        }

        insert_date_ended_iv.setColorFilter(accentColor)
        insert_description_iv.setColorFilter(accentColor)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            insert_add_btn.backgroundTintList = resources.getColorStateList(R.color.accent, null)
        } else {
            insert_add_btn.setBackgroundColor(ContextCompat.getColor(this, R.color.accent))
        }

        insert_importance_iv_1.setOnClickListener(this)
        insert_importance_iv_2.setOnClickListener(this)
        insert_importance_iv_3.setOnClickListener(this)
        insert_importance_iv_4.setOnClickListener(this)
        insert_date_started_tv.setOnClickListener(this)
        insert_date_ended_tv.setOnClickListener(this)
        insert_date_started_iv.setOnClickListener(this)
        insert_date_ended_iv.setOnClickListener(this)
        insert_description_iv.setOnClickListener(this)
        insert_add_btn.setOnClickListener(this)

    }

    private fun initWindow() {

        displayMetrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(displayMetrics)
        val width = displayMetrics!!.widthPixels
        val height = displayMetrics!!.heightPixels
        window.setLayout((width * .9).toInt(), (height * .8).toInt())

    }

    override fun displayInitCategory(category: Category) {

        insert_category_tv.text = getString(category.res)
        currentCategory = category
        drawableSelector.selectCategoryImage(insert_category_iv, category, insert_category_tv)
        drawableSelector.setImportance(
                Constants.dummyImportanceDefault,
                insert_importance_tv,
                insert_importance_iv_1,
                insert_importance_iv_2,
                insert_importance_iv_3,
                insert_importance_iv_4)

    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.insert_importance_iv_1 -> {
                drawableSelector.setSmallImportance(insert_importance_tv, insert_importance_iv_1,
                        insert_importance_iv_2, insert_importance_iv_3, insert_importance_iv_4)
                currentImportance = Importance.values()[1]
            }
            R.id.insert_importance_iv_2 -> {
                drawableSelector.setMediumImportance(insert_importance_tv, insert_importance_iv_1,
                        insert_importance_iv_2, insert_importance_iv_3, insert_importance_iv_4)
                currentImportance = Importance.values()[2]
            }
            R.id.insert_importance_iv_3 -> {
                drawableSelector.setBigImportance(insert_importance_tv, insert_importance_iv_1,
                        insert_importance_iv_2, insert_importance_iv_3, insert_importance_iv_4)
                currentImportance = Importance.values()[3]
            }
            R.id.insert_importance_iv_4 -> {
                drawableSelector.setHugeImportance(insert_importance_tv, insert_importance_iv_1,
                        insert_importance_iv_2, insert_importance_iv_3, insert_importance_iv_4)
                currentImportance = Importance.values()[4]
            }
            R.id.insert_date_started_iv -> {
                dateHelper.setDate(getString(R.string.date_started_empty), insert_date_started_tv, insert_date_ended_tv)
            }
            R.id.insert_date_ended_iv -> {
                dateHelper.setDate(getString(R.string.date_ended_empty), insert_date_started_tv, insert_date_ended_tv)
            }
            R.id.insert_date_started_tv -> {
                dateHelper.setDate(getString(R.string.date_started_empty), insert_date_started_tv, insert_date_ended_tv)
            }
            R.id.insert_date_ended_tv -> {
                dateHelper.setDate(getString(R.string.date_ended_empty), insert_date_started_tv, insert_date_ended_tv)
            }
            R.id.insert_description_iv -> setDesc()
            R.id.insert_add_btn -> insertSuccess()

            else -> {
            }
        }
    }

    private fun insertSuccess() {

        if (dateHelper.validateData(instert_title_et, insert_date_started_tv, insert_date_ended_tv)) {

            val dateStarted = dateHelper.checkBlankDate(insert_date_started_tv.text.toString())
            val dateEnded = dateHelper.checkBlankDate(insert_date_ended_tv.text.toString())

            sendData(dateStarted, dateEnded)
        }
    }

    private fun sendData(dateStarted: String, dateEnded: String) {

        val returnIntent = Intent()

        val s = SuccessModel(
                null,
                instert_title_et.text.toString(),
                currentCategory,
                insert_description_et.text.toString(),
                dateStarted, dateEnded,
                currentImportance)
        //s.setId(UUID.randomUUID().toString());
        //Log.e(TAG, "sendData: uuid" + s.getId() );

        returnIntent.putExtra(Constants.EXTRA_INSERT_SUCCESS_ITEM, s)
        setResult(Activity.RESULT_OK, returnIntent)
        finish()
    }


    private fun setDesc() {

        val intent = Intent(this@InsertSuccessActivity, EditDescriptionActivity::class.java)
        intent.putExtra("description", insert_description_et.text.toString())
        startActivityForResult(intent, Constants.REQUEST_CODE_DESCRIPTION)

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == Constants.REQUEST_CODE_DESCRIPTION) {
            if (resultCode == Activity.RESULT_OK) {
                val description = data.extras.getString(Constants.EXTRA_DESCRIPTION)
                insert_description_et.setText(description)

            }
        }
    }

}
